### PR TITLE
Send code coverage reports to CodeCov

### LINKED ISSUE #6101: Integrate CodeCov
Continue of #6048 

This PR is about integrating CodeCov tool to help us track code coverage.

### PR DESCRIPTION
**Pull request by wzieba**
Closes: #6101 

### Description

This PR adds integration of CodeCov tool.

### Testing instructions

You can navigate to https://app.codecov.io/gh/woocommerce/woocommerce-android and browse statistics but keep in mind that there's no yet report for `trunk` so you might experience some 404s or other strange behaviour (CodeCov has nothing to compare with).

Also this PR didn't change any code that is meant to be covered. For testing purposes I've added (and later removed) some test commit. Please see it here: https://app.codecov.io/gh/woocommerce/woocommerce-android/commit/96f3e62278c56968c34ffe008ac8a39bdc67b986

### Images/gif

image

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -57,3 +57,12 @@ steps:
     command: ".buildkite/commands/installable-build.sh"
     if: "build.pull_request.id != null || build.pull_request.draft"
     plugins: *common_plugins
+
+  - label: "Generate code coverage and send it to CodeCov"
```
Comment (wzieba) [inline .buildkite/pipeline.yml]: ~I decided to go with a separate step rather than uploading report after "Test WooCommerce" and "Test CardReaderModule" because:~
1. ~We need a consolidated report of two of those tasks~
2. ~They run on `vanillaRelease` variant which produces different results than dev-first `wasabiDebug` (due to code shrinking).~

| wasabiDebug | vanillaRelease |
| --- | --- |
| image | image |

Update:

in #6125 we start to use `jalapeno` for every step so this issue is not present anymore

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -57,3 +57,12 @@ steps:
...
+    command: |
+      echo "--- ⚒️ Testing debug & generating code coverage"
+      cp gradle.properties-example gradle.properties
+      ./gradlew jacocoTestReport
+      curl -Os https://uploader.codecov.io/latest/linux/codecov
+      chmod +x codecov
+      ./codecov -t $CODECOV_TOKEN
```
Comment (wzieba) [inline .buildkite/pipeline.yml]: I tried to add [integrity check](https://docs.codecov.com/docs/codecov-uploader#integrity-checking-the-uploader) but failed as Amazon Linux 2 image seems to not have required dependencies: `/bin/bash: line 8: shasum: command not found` (see 24dfce5).

1. Do you think we need integrity check? I see that e.g. [WooCommerce](https://github.com/woocommerce/woocommerce/blob/trunk/.github/workflows/pr-code-coverage.yml#L71)  doesn't have it - do you think it'd be needed for Woo Android?
2. If yes - do you have an idea how to resolve lack of `shasum` ?

File: codecov.yml [hunk 3]
```diff
@@ -0,0 +1 @@
+comment: false
```
Comment (wzieba) [inline codecov.yml]: This disables comments on every single PR which, at least for now, is not very useful feature and might to introduce some chaos in PRs.

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (wzieba) [reply .buildkite/pipeline.yml]: Internal discussion about used variants: p91TBi-7YK-p2

**Comment by wzieba** [comment]
Hi @jkmassel - could you please take a look if this change looks okay to you? I'm the most concerned with integrity check (https://github.com/woocommerce/woocommerce-android/pull/6106#discussion_r836595774), WDYT?

**Comment by wzieba** [comment]
Hi @ParaskP7 👋 . I see you've been here 😄 do you mind reviewing this PR? I'm currently on the HACK Week so I can work on the integration - but I'm not sure if the integrity check is a must-have for us?

**Comment by ParaskP7** [comment]
👋 @wzieba !

Thanks for the ping!

> Hi @ParaskP7 👋 . I see you've been here 😄 do you mind reviewing this PR?

Definitely, will do! 💯 

PS: I was actually eyeballing this PR, but since I saw you pinging @jkmassel directly I thought you would like to get an initial feedback from him first.

> I'm currently on the HACK Week so I can work on the integration - but I'm not sure if the integrity check is a must-have for us?

👍

File: .buildkite/pipeline.yml (hunk 2 above)
Comment (ParaskP7) [reply .buildkite/pipeline.yml]: I personally have little knowledge about all that, maybe @jkmassel can help. 🤔

File: codecov.yml (hunk 3 above)
Comment (ParaskP7) [reply codecov.yml]: True, but I've seen PRs from other projects that was also using this and was actually looking good, at least the way I saw that. I would have thought that this is the main benefit of `CodeCov`. How else would engineers utilize this tools in your opinion? 🤔

File: .buildkite/pipeline.yml [hunk 4]
```diff
@@ -38,6 +38,11 @@ steps:
       cp gradle.properties-example gradle.properties
       ./gradlew testJalapenoDebug
       ./gradlew lib:cardreader:testDebug
+      echo "--- ⚒️ Generating and uploading code coverage"
+      ./gradlew jacocoTestReport
+      curl -Os https://uploader.codecov.io/latest/linux/codecov
+      chmod +x codecov
+      ./codecov -t $CODECOV_TOKEN
```
Comment (ParaskP7) [inline .buildkite/pipeline.yml]: Question (❓): Initially I saw you adding the `CodeCov` step as a separate step. Then, you decided to move it as part of the `Unit Tests` step. Why did you chose to do so? 🤔

File: config/gradle/jacoco.gradle [hunk 5]
```diff
@@ -8,10 +8,10 @@ jacoco {
 
 rootProject.afterEvaluate {
     tasks.register("jacocoTestReport", JacocoReport) {
-        dependsOn(':WooCommerce:testWasabiDebugUnitTest', ':libs:cardreader:testDebugUnitTest')
+        dependsOn(':WooCommerce:testJalapenoDebugUnitTest', ':libs:cardreader:testDebugUnitTest')
```
Comment (ParaskP7) [inline config/gradle/jacoco.gradle:11]: Question (❓): Why did you change `wasabi` with `jalapeno`, is it because ultimately this task is going to be mainly used via `CI`? 🤔

**Review by ParaskP7 — COMMENTED** [review]
LGTM! 💯 

I have left a couple of questions (❓), but unfortunately I would still depend on @jkmassel for an answer on the integrity check. Thus, I am going to just comment on this, without approving it just yet, but it is anyway a soft 👍 and a 🟢 light from my side.

File: config/gradle/jacoco.gradle (hunk 5 above)
Comment (wzieba) [reply config/gradle/jacoco.gradle:11]: Yes exactly as you write - I believe that the vast majority of times (if not always) CI-usage will be the only one.

File: .buildkite/pipeline.yml (hunk 4 above)
Comment (wzieba) [reply .buildkite/pipeline.yml]: Good point! I realized that steps in Buildkite seem to be executed in different environments as in every execution we start with preparing the `gradle.properties` file.

Generating JaCoCo artifacts requires running tests. This means that handling CodeCov in a separate step would make us run tests twice.

File: codecov.yml (hunk 3 above)
Comment (wzieba) [reply codecov.yml]: I was thinking that we can use it as a metric and then decide which areas of the app need our attention with coverage. But I'm okay with using PR comments - we can always disable them later if they'll be problematic.

File: config/gradle/jacoco.gradle (hunk 5 above)
Comment (ParaskP7) [reply config/gradle/jacoco.gradle:11]: Thanks for the clarification @wzieba ! 🙇

File: .buildkite/pipeline.yml (hunk 4 above)
Comment (ParaskP7) [reply .buildkite/pipeline.yml]: Make sense, again thanks for the clarification @wzieba ! 🙇 

PS: Let's also monitor how much the build time increases because of this extra addition. I recently heard a [podcast](https://talkingkotlin.com/slacking-with-zac-sweers/) that Slack removed code coverage on CI and they won back 10' mins of build time. 🤷

File: codecov.yml (hunk 3 above)
Comment (ParaskP7) [reply codecov.yml]: Indeed, we can start with them and then listen to feedback and act on it accordingly. Instead, if we start with it disabled we might never got back to try it afterwards, forgetting about it. It seems to me that this is such a cool feature to not try at least... 😅

Comment (wzieba) [reply codecov.yml]: Done in 6d32b77 !

File: .buildkite/pipeline.yml (hunk 4 above)
Comment (wzieba) [reply .buildkite/pipeline.yml]: Sure, good point. We can precisely track how much time this task takes on CI using [tracks-gradle](https://github.com/wzieba/tracks-gradle) 😄.

Comment (ParaskP7) [reply .buildkite/pipeline.yml]: > ...using [tracks-gradle](https://github.com/wzieba/tracks-gradle) 😄.

Yes sir! 👏

File: codecov.yml (hunk 3 above)
Comment (ParaskP7) [reply codecov.yml]: Awesome, now let's check on the PR comment that `CodeCov` generates and take it from there! 🕺

File: .buildkite/pipeline.yml (hunk 2 above)
Comment (wzieba) [reply .buildkite/pipeline.yml]: I rethought the issue and I think it's a good idea to add an integrity check. We do this e.g. for the dependencies diff step. I'll prepare a Platform9 request for installing `shasum` on those machines and I believe we'll be unblocked 🙂.

Comment (wzieba) [reply .buildkite/pipeline.yml]: Internal communication link: pdnsEh-fe-p2

**Comment by wzieba** [comment]
@ParaskP7 integrity check has been added. I think we're ready for merge, WDYT?

File: codecov.yml [hunk 6]
```diff
@@ -0,0 +1,3 @@
+comment:
+  require_changes: true
```
Comment (wzieba) [inline codecov.yml:2]: Docs: https://docs.codecov.com/docs/pull-request-comments#requiring-changes

TLDR; post comment only when there's a change in code coverage. This is for e.g. dependabot PRs.

File: .buildkite/pipeline.yml [hunk 7]
```diff
@@ -38,6 +38,16 @@ steps:
...
+      echo "--- ⚒️ Generating and uploading code coverage"
+      ./gradlew jacocoTestReport
+      curl https://keybase.io/codecovsecurity/pgp_keys.asc | gpg --no-default-keyring --keyring trustedkeys.gpg --import
+      curl -Os https://uploader.codecov.io/latest/linux/codecov
+      curl -Os https://uploader.codecov.io/latest/linux/codecov.SHA256SUM
+      curl -Os https://uploader.codecov.io/latest/linux/codecov.SHA256SUM.sig
+      gpgv codecov.SHA256SUM.sig codecov.SHA256SUM
+      sha256sum -c codecov.SHA256SUM
+      chmod +x codecov
+      ./codecov -t $CODECOV_TOKEN
```
Comment (AliSoftware) [inline .buildkite/pipeline.yml]: May I suggest moving all those lines into à `.buildkite/commands/run_jacoco.sh` file then just invoke that script from that part of the pipeline, instead of having all those related lines being inlined in the pipeline?

(Similar to how it has been done for installable build step below)

This is obviously not a big deal  nor a blocker, but would:
- make the pipeline more readable
- allow to isolate so many related lines in a dedicated place (just my personal opinion but imho it's fine to keep commands inlined if it's only 2-3 simple commands, but when it gets more complex than that it's better to move it in a script)
- and would make it simpler to move that whole logic into our `bash-cache` plugin in the future if more projects start to want to adopt it too and we want to generalize the logic

😊

Comment (wzieba) [reply .buildkite/pipeline.yml]: Good suggestion, thanks! Addressed in 80342b3 and a865114.

File: .buildkite/commands/upload-code-coverage.sh [hunk 8]
```diff
@@ -0,0 +1,8 @@
+curl https://keybase.io/codecovsecurity/pgp_keys.asc | gpg --no-default-keyring --keyring trustedkeys.gpg --import
```
Comment (AliSoftware) [inline .buildkite/commands/upload-code-coverage.sh:3]: Can you please add a shebang (i.e. `#!/bin/bash -eu`) to this script file? 🙇‍♂️

**Review by ParaskP7 — APPROVED** [review]
👋 @wzieba !

It is a 👍 from my side and good job, I'll wait for someone else to double-approve on the integrity check, just to be on the safe side, but it's a 🟢 light from me.

🚀

File: .buildkite/commands/upload-code-coverage.sh [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+curl -Os https://uploader.codecov.io/latest/linux/codecov
+curl -Os https://uploader.codecov.io/latest/linux/codecov.SHA256SUM
+curl -Os https://uploader.codecov.io/latest/linux/codecov.SHA256SUM.sig
+gpgv codecov.SHA256SUM.sig codecov.SHA256SUM
+sha256sum -c codecov.SHA256SUM
+chmod +x codecov
+./codecov -t $CODECOV_TOKEN
```
Comment (AliSoftware) [inline .buildkite/commands/upload-code-coverage.sh]: Running [`shellcheck`](https://github.com/koalaman/shellcheck) on this script to lint it suggested me to secure this variable expansion by putting it in quotes:
```
$ shellcheck .buildkite/commands/*            

In .buildkite/commands/upload-code-coverage.sh line 1:
curl https://keybase.io/codecovsecurity/pgp_keys.asc | gpg --no-default-keyring --keyring trustedkeys.gpg --import
^-- SC2148 (error): Tips depend on target shell and yours is unknown. Add a shebang or a 'shell' directive.


In .buildkite/commands/upload-code-coverage.sh line 8:
./codecov -t $CODECOV_TOKEN
             ^------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
./codecov -t "$CODECOV_TOKEN"

For more information:
  https://www.shellcheck.net/wiki/SC2148 -- Tips depend on target shell and y...
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

```suggestion
./codecov -t "$CODECOV_TOKEN"
```

File: .buildkite/commands/upload-code-coverage.sh (hunk 8 above)
Comment (AliSoftware) [reply .buildkite/commands/upload-code-coverage.sh:3]:
```suggestion
#!/bin/bash -eu

curl https://keybase.io/codecovsecurity/pgp_keys.asc | gpg --no-default-keyring --keyring trustedkeys.gpg --import
```

File: .buildkite/commands/upload-code-coverage.sh (hunk 9 above)
Comment (wzieba) [reply .buildkite/commands/upload-code-coverage.sh]: Thanks! TIL about `shellcheck`

**Comment by wzieba** [comment]
> Good occasion to add it to .gitignore to avoid similar accident in the future?

I'd opt to not add `.gitignore` as this script is never meant to be executed on developer machines (and it can't as they don't have `CODECOV_TOKEN` so adding this be confusing or/and redundant. But it's not a strong opinion, if you'd still prefer adding `.gitignore`, I'll do this. WDYT @AliSoftware ?

**Review by AliSoftware — APPROVED** [review]
> I'd opt to not add .gitignore as this script is never meant to be executed on developer machines (and it can't as they don't have CODECOV_TOKEN so adding this be confusing or/and redundant. But it's not a strong opinion, if you'd still prefer adding .gitignore, I'll do this. WDYT @AliSoftware ?

Good point. Tbh while I thought about this (only being supposed to be run by CI) when I wrote that suggestion and though, like you, that it might not be strictly necessary, I though "but there could still be a case when a developer or an MIE would still run `.buildkite/commands/upload_code_coverage.sh` manually on their own machine, like to debug stuff as Platform Engineers or whatnot, so…"

But given how seldom that would be, and that the script would fail for a likely lack of `CODECOV_TOKEN` being defined even on an Infra Engineer's machine like me, and that we'd have to also ignore the shasum and signature files, … yeah. not worth it and makes sense not to add it to gitignore after all.

---

All good for me then! :shipit:
